File: P01/Backend/Python/utils/inventario.py
class Producto:
    def __init__(self, p_nombre: str, p_cantidad: int, p_precio: float):
        self.p_nombre = p_nombre
        self.p_cantidad = p_cantidad
        self.p_precio = p_precio

    def retorno_producto(self):
        return { "p_nombre": self.p_nombre, "p_cantidad": self.p_cantidad, "p_precio": self.p_precio }

    def __repr__(self):
        return f"Producto(nombre={self.p_nombre}, cantidad={self.p_cantidad}, precio={self.p_precio})"


class Inventario:
    def __init__(self):
        self.productos = []

    def nuevo_producto(self, p_nombre: str, p_cantidad: int, p_precio: float):
        if any(producto.p_nombre == p_nombre for producto in self.productos):
            raise ValueError(f"Producto '{p_nombre}' ya existente")
        
        if not p_nombre or p_cantidad < 0 or p_precio < 0:
            raise ValueError("Ingreso de valores no validos")
        
        self.productos.append(Producto(p_nombre, p_cantidad, p_precio))

    def eliminar_producto(self, p_nombre: str) -> bool:
        initial_length = len(self.productos)
        self.productos = [p for p in self.productos if p.p_nombre != p_nombre]
        return len(self.productos) < initial_length

    def buscar_producto(self, p_nombre: str):
        for producto in self.productos:
            if producto.p_nombre == p_nombre:
                return producto
        return None

    def ordenar_producto(self, precio=False, cantidad=False):
        if precio:
            return sorted(self.productos, key=lambda p: p.p_precio)
        elif cantidad:
            return sorted(self.productos, key=lambda p: p.p_cantidad)
        return self.productos

File: P01/Backend/Node/utils/inventario.js
class Producto {
    constructor(p_nombre, p_cantidad, p_precio) {
        this.p_nombre = p_nombre;
        this.p_cantidad = p_cantidad;
        this.p_precio = p_precio;
    }

    retornoProducto() {
        return {
            p_nombre: this.p_nombre,
            p_cantidad: this.p_cantidad,
            p_precio: this.p_precio
        };
    }

    toString() {
        return `Producto(nombre=${this.p_nombre}, cantidad=${this.p_cantidad}, precio=${this.p_precio})`;
    }
}

class Inventario {
    constructor() {
        this.productos = [];
    }

    nuevoProducto(p_nombre, p_cantidad, p_precio) {
        if (this.productos.some(producto => producto.p_nombre === p_nombre)) {
            throw new Error(`Producto '${p_nombre}' ya existente`);
        }

        if (!p_nombre || p_cantidad < 0 || p_precio < 0) {
            throw new Error("Ingreso de valores no válidos");
        }

        this.productos.push(new Producto(p_nombre, p_cantidad, p_precio));
    }

    eliminarProducto(p_nombre) {
        const initialLength = this.productos.length;
        this.productos = this.productos.filter(producto => producto.p_nombre !== p_nombre);
        return this.productos.length < initialLength;
    }

    buscarProducto(p_nombre) {
        return this.productos.find(producto => producto.p_nombre === p_nombre) || null;
    }

    ordenarProducto(precio = false, cantidad = false) {
        if (precio) {
            return this.productos.sort((a, b) => a.p_precio - b.p_precio);
        } else if (cantidad) {
            return this.productos.sort((a, b) => a.p_cantidad - b.p_cantidad);
        }
        return this.productos;
    }

    listarProductos() {
        return this.productos;
    }
}

module.exports = { Inventario };